helix-rest/src/main/java/org/apache/helix/rest/server/service/InstanceServiceImpl.java

```diff
@@ -122,115 +122,166 @@ public InstanceInfo getInstanceInfo(String clusterId, String instanceName,
-      StoppableCheck result =
-          new StoppableCheck(_customRestClient.getInstanceStoppableCheck(baseUrl, customPayLoads),
-              StoppableCheck.Category.CUSTOM_INSTANCE_CHECK);
-      if (!result.isStoppable()) {
-        return result;
-      }
-      LOG.info("{} passed client side instance level health checks", instanceName);
-      return performPartitionLevelChecks(clusterId, instanceName, baseUrl, customPayLoads);
+      return new StoppableCheck(_customRestClient.getInstanceStoppableCheck(baseUrl, customPayLoads),
+          StoppableCheck.Category.CUSTOM_INSTANCE_CHECK);
     } catch (IOException e) {
-      LOG.error("Failed to perform custom client side instance level health checks for {}/{}",
-          clusterId, instanceName, e);
+      LOG.error("Failed to perform custom client side instance level health checks for {}/{}", clusterId, instanceName,
+          e);
       throw e;
     }
   }
 
-  @VisibleForTesting
-  protected StoppableCheck performPartitionLevelChecks(String clusterId, String instanceName,
-      String baseUrl, Map<String, String> customPayLoads) throws IOException {
-    LOG.info("Perform partition level health checks for {}/{}", clusterId, instanceName);
-    // pull the health status from ZK
+  private Map<String, StoppableCheck> performPartitionsCheck(List<String> instances, RESTConfig restConfig,
+      Map<String, String> customPayLoads) {
     PartitionHealth clusterPartitionsHealth = generatePartitionHealthMapFromZK();
-    Map<String, List<String>> expiredPartitionsOnInstances =
-        clusterPartitionsHealth.getExpiredRecords();
     // update the health status for those expired partitions on instances
-    try {
-      for (Map.Entry<String, List<String>> entry : expiredPartitionsOnInstances.entrySet()) {
-        Map<String, Boolean> partitionHealthStatus =
-            _customRestClient.getPartitionStoppableCheck(baseUrl, entry.getValue(), customPayLoads);
-        partitionHealthStatus.entrySet().forEach(kv -> clusterPartitionsHealth
-            .updatePartitionHealth(instanceName, kv.getKey(), kv.getValue()));
-      }
-    } catch (IOException e) {
-      LOG.error("Failed to perform client side partition level health checks for {}/{}", clusterId,
-          instanceName, e);
-      throw e;
+    Map<String, List<String>> expiredPartitionsByInstance = clusterPartitionsHealth.getExpiredRecords();
+    Map<String, Future<Map<String, Boolean>>> updatedPartitionsHealthStatusByInstance = new HashMap<>();
+    for (Map.Entry<String, List<String>> entry : expiredPartitionsByInstance.entrySet()) {
+      String instance = entry.getKey();
+      List<String> expiredPartitions = entry.getValue();
+      Callable<Map<String, Boolean>> refreshTask =
+          () -> _customRestClient.getPartitionStoppableCheck(getBaseUrl(instance, restConfig), expiredPartitions,
+              customPayLoads);
+      updatedPartitionsHealthStatusByInstance.put(instance, POOL.submit(refreshTask));
```

Comment:
I have a question here. If the tasks are submitted, will that be guaranteed that the next line of code will be executed after all tasks are finished? Because next step will be final stoppable result update.

Reply:
I don't understand why do you think the next code will get skipped?

Reply:
thread pool submission is async operation. Not sure whether Java 8 is doing same thing for POOL.submit.

Reply:
I think the behavior is still the same, we'll find out in later unit tests